File: NFCTimeSheets/NFCTimeSheets/TapInbox.swift
//
//  TapInbox.swift
//  NFCTimeSheets
//
//  The hand-off between "iOS opened a universal link off a tag" and "the Log tab wrote
//  a row". Lifted out of NFCTimeSheetsApp.swift so it is Foundation + Observation only:
//  the cold-launch ordering below is the exact thing that lost the owner's first real
//  tap, and it is now pinned by checks/tap-inbox-check.swift outside Xcode.
//

import Foundation
import Observation

/// Where a tap arrives from, whichever way it came in: the in-app NDEF scan, or iOS
/// opening the universal link off the tag while the app was in the background.
///
/// Both paths can fire for a single physical tap, so identical taps inside a short
/// window collapse into one. Without this, one tap would clock in and straight back out.
///
/// COLD LAUNCH ORDERING - the reason this is a mailbox and not a callback.
/// `onOpenURL` fires in NFCTimeSheetsApp while `Session.restore()` is still running, so
/// on a tap-launch the tap arrives BEFORE LogView exists. Two orderings have to work and
/// both are covered by the consumer in LogView:
///
///   set-before-mount: accept() parks the id here while ContentView is still showing the
///     launch spinner. LogView mounts later and `.task` calls take() -> handled once.
///     `.onChange` does NOT fire for a value that was already set when the view appeared
///     (no `initial: true`), so it cannot handle the same tap a second time.
///   set-after-mount: LogView is already up, `.task` already took nil. accept() flips
///     pendingLocationId nil -> X, `.onChange` fires, take() returns X -> handled once.
///
/// In both orderings take() then flips X -> nil, which fires `.onChange` again with a
/// nil id; the consumer's `guard id != nil` drops that echo. So: never lost, never twice.
@Observable
final class TapInbox {
    var pendingLocationId: String?
    private var last: (locationId: String, at: Date)?
    private let window: TimeInterval = 3

    func accept(_ locationId: String) {
        if let last, last.locationId == locationId, Date.now.timeIntervalSince(last.at) < window { return }
        last = (locationId, .now)
        pendingLocationId = locationId
    }

    func take() -> String? {
        defer { pendingLocationId = nil }
        return pendingLocationId
    }
}

File: android/app/src/main/kotlin/io/github/qwadratic/nfctimesheets/core/TapInbox.kt
package io.github.qwadratic.nfctimesheets.core

/**
 * The hand-off between "Android opened an App Link off a tag" and "the log screen wrote
 * a row". Port of NFCTimeSheets/NFCTimeSheets/TapInbox.swift, including the reason it
 * exists — the cold-launch ordering below is the exact thing that lost the owner's first
 * real tap on iOS, and it is not going to be re-discovered here.
 *
 * A tap can arrive from three places and all of them land in this one mailbox:
 *   - Android 16+  : ACTION_VIEW App Link, delivered to MainActivity.onCreate/onNewIntent
 *   - Android <=15 : ACTION_NDEF_DISCOVERED, delivered to NfcTapActivity, forwarded here
 *   - the user     : opening the link from a browser or a message
 *
 * COLD LAUNCH ORDERING — the reason this is a mailbox and not a callback.
 * `onCreate` sees the intent while the session is still being restored from the server,
 * so on a tap-launch the tap arrives BEFORE any screen that could act on it exists.
 * Two orderings have to work:
 *
 *   set-before-consumer: accept() parks the id while the UI is still a spinner. The log
 *     screen appears later and takes it -> handled once.
 *   set-after-consumer: the screen is already up and observing; accept() flips pending
 *     null -> X, the observer takes X -> handled once.
 *
 * In both orderings take() then flips X -> null; a consumer that guards on non-null
 * drops that echo. So: never lost, never twice.
 *
 * Not a StateFlow and not Android-aware, so android/checks can run the orderings.
 */
class TapInbox(private val nowMillis: () -> Long = System::currentTimeMillis) {

    var pendingLocationId: String? = null
        private set

    private var lastLocationId: String? = null
    private var lastAtMillis: Long = 0

    /**
     * One physical tap can be delivered twice (the NDEF dispatch AND the App Link), so
     * identical taps inside a short window collapse into one. Without this, one tap
     * would clock in and straight back out.
     */
    fun accept(locationId: String): Boolean {
        val now = nowMillis()
        if (locationId == lastLocationId && now - lastAtMillis < WINDOW_MS) return false
        lastLocationId = locationId
        lastAtMillis = now
        pendingLocationId = locationId
        return true
    }

    fun take(): String? {
        val id = pendingLocationId
        pendingLocationId = null
        return id
    }

    private companion object {
        const val WINDOW_MS = 3_000L
    }
}
